xtrlock.man
.TH XTRLOCK 1
.SH NAME
xtrlock \- Lock X display until password supplied, leaving windows visible
.SH SYNOPSIS
.B xtrlock [-b] [-c command] [-f]
.SH DESCRIPTION
.B xtrlock
locks the X server till the user enters their password at the keyboard.

While
.B xtrlock
is running, the mouse and keyboard are grabbed and the mouse cursor
becomes a padlock.  Output displayed by X programs, and windows put up
by new X clients, continue to be visible, and any new output is
displayed normally.

The mouse and keyboard are returned when the user types their
password, followed by Enter or Newline.  If an incorrect password is
entered the bell is sounded.  Pressing Backspace or Delete erases one
character of a password partially typed; pressing Escape or Clear
clears anything that has been entered.

If too many attempts are made in too short a time further keystrokes
generate bells and are otherwise ignored until a timeout has expired.

The X server screen saver continues to operate normally; if it comes
into operation the display may be restored by the usual means of
touching a key (Shift, for example) or the mouse.
.SH OPTIONS
.TP
\fB\-b\fR
blank the screen as well as displaying the padlock
.TP
\fB\-c\fR command
run specified command after unlocking screen
.TP
\fB\-f\fR
fork after locking is complete, and return success from the parent
process
.SH X RESOURCES, CONFIGURATION
None.
.SH BUGS
Additional input devices other than the keyboard and mouse are not
disabled.

The timeouts, bitmaps and mouse cursor colour cannot be modifed.
.SH SEE ALSO
.BR X "(1), Xlib Documentation."
.SH AUTHOR
Ian Jackson <ian@chiark.greenend.org.uk>

Unofficial fork to piggyback command by Arun Prakash Jana <engineerarun@gmail.com>
.SH COPYRIGHT
Copyright (C) 1993, 1994 by Ian Jackson.

Permission to use, copy, modify and distribute this software and its
documentation is granted under the terms of the GNU General Public
Licence, version 2 or later, as published by the Free Software
Foundation.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.


debian/changelog
xtrlock (2.8) unstable; urgency=low

  * patch from Simon Tatham to add a -f option [fork, and return success
    from parent once lock acquired] (Closes: #823685)

 -- Matthew Vernon <matthew@debian.org>  Sat, 21 May 2016 19:08:12 +0100

xtrlock (2.7) unstable; urgency=low

  * patch from Chris Lamb to make builds reproducible by adding -n to gzip
    invocations (Closes: #777351)

 -- Matthew Vernon <matthew@debian.org>  Sat, 07 Mar 2015 22:34:45 +0000

xtrlock (2.6) unstable; urgency=low

  * patch from Simon Ruderich to correctly add hardening flags
  * check returns of setuid and setgid
  * build-depend on dpkg-dev
  
 -- Matthew Vernon <matthew@debian.org>  Tue, 07 Jan 2014 14:22:55 +0000

xtrlock (2.5) unstable; urgency=low

  * correct contributor name (sorry!)
  * add hardening flags
  * update version number

 -- Matthew Vernon <matthew@debian.org>  Sun, 05 Jan 2014 16:27:52 +0000

xtrlock (2.4) unstable; urgency=low

  * add a -b option to blank the screen (code based on patch by Markus,
    waldeck@gmx.de, for which thanks; documentation by myself. 
    (Closes: #636993)

 -- Matthew Vernon <matthew@debian.org>  Sun, 05 Jan 2014 12:01:48 +0000

xtrlock (2.3) unstable; urgency=low

  * remove reference to xlock, which is no longer in Debian (Closes: #721724)

 -- Matthew Vernon <matthew@debian.org>  Wed, 04 Sep 2013 15:37:38 +0100

xtrlock (2.2) unstable; urgency=low

  * Move menu entry to /usr/share/menu
  * Change menu section to Screen/Locking
  * Update standards-version
  * Add a lintian override for being setgid shadow
  
 -- Matthew Vernon <matthew@debian.org>  Tue, 26 Jun 2012 20:22:20 +0100

xtrlock (2.1) unstable; urgency=low

  * never exit non-zero without saying why.
  * try repeatedly, with waits, to grab the keyboard (Closes: #316343).
  * include a copy of the GPL in the source package (Closes: #519327).
  * honour nostrip DEB_BUILD_OPTION (Closes: #438346).
  * Debian's package is essentially upstream; only include one changelog
    in the binary package (rather than one and a symlink)
  * Don't install README.Debian any more, as it's mostly just build
    instructions
  
 -- Matthew Vernon <matthew@debian.org>  Tue, 26 Jun 2012 19:12:02 +0100

xtrlock (2.0-14) unstable; urgency=low

  * remove obsolete build-depends on xutils (closes: #579036)

 -- Matthew Vernon <matthew@debian.org>  Thu, 13 May 2010 18:01:51 +0100

xtrlock (2.0-13) unstable; urgency=low

  * build-depend on x11proto-core-dev rather than the old x-dev
  * build-depend on xutils-dev (closes: #485728)

 -- Matthew Vernon <matthew@debian.org>  Tue, 23 Dec 2008 11:36:04 +0000

xtrlock (2.0-12) unstable; urgency=low

  * Install into /usr not /usr/X11R6 (closes: #362206, #363293)

 -- Matthew Vernon <matthew@debian.org>  Tue, 18 Apr 2006 17:19:10 +0100

xtrlock (2.0-11) unstable; urgency=low
  * Fix build-depends: for new etch (closes: #346840)

 -- Matthew Vernon <matthew@debian.org>  Sat, 14 Jan 2006 15:30:28 +0000
	
xtrlock (2.0-10) unstable; urgency=low

  * Take over this package, since I seem to have some tuits
  * Correct the Author's address in the manpage to something sensible
  * Similarly the original packager's.
  * For reference, the security problem fixed in -9 was CAN-2005-0079
	
 -- Matthew Vernon <matthew@debian.org>  Thu, 20 Jan 2005 14:02:10 +0000
	
xtrlock (2.0-9) unstable; urgency=high

  * Fix the problem whereby we unlocked on long input (closes: #278191, #278190)
  * tidy up a switch statement (closes: #264173)

 -- Matthew Vernon <matthew@debian.org>  Mon, 17 Jan 2005 10:47:09 +0000
	
xtrlock (2.0-8) unstable; urgency=low

  * Uploading with maintainer set to QA group

 -- Andrew Pollock <apollock@debian.org>  Sun, 15 Feb 2004 18:32:12 +1100

xtrlock (2.0-7) unstable; urgency=low

  * Maintainer upload.
  * Update to standards version 3.5.6.
  * Removes buffer overrun potentiality (closes: #154738)
  * Drop setgid privileges (closes: #154740)

 -- Martin Mitchell <martin@debian.org>  Sun,  1 Sep 2002 14:49:26 +1000

xtrlock (2.0-6.1) unstable; urgency=low

  * NMU
  * Removed the buffer overrun potentiallity (closes: #154738)
  * Update the code for remove the possibilty to read shadow file 
  (closes: #154740)
  * Fix spelling error in the copyright file
  * Fix the place where is stored the common-licenses 
  * Update the debian/rules to:
    - strip the /usr/X11R6/bin/xtrlock
    - include the section and priority in control
  * Update the debian/control file:
    - include the build-depends
    - update the standard-version to 3.5.6

 -- Otavio Salvador <otavio@debian.org>  Sat, 31 Aug 2002 16:23:21 -0300

xtrlock (2.0-6) unstable; urgency=low

  * Update to standards version 3.0.1.
  * Add menu entry. (closes: #46201)

 -- Martin Mitchell <martin@debian.org>  Thu, 28 Oct 1999 18:38:14 +1000

xtrlock (2.0-5) unstable; urgency=low

  * Correct installation path during package build. (#25055)
  * Update to standards version 2.4.1.
  * Update copyright file.

 -- Martin Mitchell <martin@debian.org>  Sun,  2 Aug 1998 13:01:29 +1000

xtrlock (2.0-4) unstable; urgency=low

  * Make xtrlock setgid shadow, instead of setuid root. (#7635,#8543)

 -- Martin Mitchell <martin@debian.org>  Thu,  4 Dec 1997 00:27:07 +1100

xtrlock (2.0-3) unstable; urgency=low

  * New maintainer.

 -- Martin Mitchell <martin@debian.org>  Thu, 27 Nov 1997 01:34:14 +1100

xtrlock (2.0-2.2) unstable; urgency=low
  
  * Non-maintainer release.
  * Libc6 compile.
  * Update from pre-2.0.0.0 standards.

 -- Martin Mitchell <martin@debian.org>  Wed, 22 Oct 1997 18:07:31 +1000

xtrlock-2.0 Debian 3 - Michael Meskes <meskes@debian.org>
Include shadow patches
Add architecture field (Bug#4041)
Minor changes to debian.rules
gzip manpage

xtrlock-2.0 Debian 0 - sde

Initial release.





xtrlock.c
/*
 * xtrlock.c
 *
 * X Transparent Lock
 *
 * Copyright (C)1993,1994 Ian Jackson
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include <X11/X.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/keysym.h>
#include <X11/Xos.h>

#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <errno.h>
#include <pwd.h>
#include <grp.h>
#include <limits.h>
#include <string.h>
#include <crypt.h>
#include <unistd.h>
#include <math.h>
#include <ctype.h>
#include <values.h>

#ifdef SHADOW_PWD
#include <shadow.h>
#endif

#include "lock.bitmap"
#include "mask.bitmap"
#include "patchlevel.h"

Display *display;
Window window, root;

#define TIMEOUTPERATTEMPT 30000
#define MAXGOODWILL  (TIMEOUTPERATTEMPT*5)
#define INITIALGOODWILL MAXGOODWILL
#define GOODWILLPORTION 0.3
#define CMDMAXLEN 64

struct passwd *pw;
int passwordok(const char *s) {
#if 0
  char key[3];
  char *encr;

  key[0] = *(pw->pw_passwd);
  key[1] =  (pw->pw_passwd)[1];
  key[2] =  0;
  encr = crypt(s, key);
  return !strcmp(encr, pw->pw_passwd);
#else
  /* simpler, and should work with crypt() algorithms using longer
     salt strings (like the md5-based one on freebsd).  --marekm */
  return !strcmp(crypt(s, pw->pw_passwd), pw->pw_passwd);
#endif
}

int main(int argc, char **argv){
  XEvent ev;
  KeySym ks;
  char cbuf[10], rbuf[128]; /* shadow appears to suggest 127 a good value here */
  int clen, rlen=0;
  long goodwill= INITIALGOODWILL, timeout= 0;
  XSetWindowAttributes attrib;
  Cursor cursor;
  Pixmap csr_source,csr_mask;
  XColor csr_fg, csr_bg, dummy, black;
  int ret, screen, blank = 0, fork_after = 0;
#ifdef SHADOW_PWD
  struct spwd *sp;
#endif
  struct timeval tv;
  int tvt, gs;
  int opt, index, cmdlen = 0;
  char cmdstr[CMDMAXLEN] = {0};

#if 0
  while (argc > 1) {
    if ((strcmp(argv[1], "-b") == 0)) {
      blank = 1;
      argc--;
      argv++;
    } else if ((strcmp(argv[1], "-f") == 0)) {
      fork_after = 1;
      argc--;
      argv++;
    } else {
      fprintf(stderr,"xtrlock (version %s); usage: xtrlock [-b] [-f]\n",
              program_version);
      exit(1);
    }
  }
#endif

  while ((opt = getopt (argc, argv, "bc:f")) != -1) {
    switch (opt) {
      case 'b':
        blank = 1;
        break;
      case 'c':
        strncpy(cmdstr, optarg, CMDMAXLEN - 1);
        cmdlen = strlen(cmdstr);
        break;
      case 'f':
        fork_after = 1;
        break;
      default:
        fprintf(stderr,"xtrlock (version %s mod); usage: xtrlock [-b] [-c command] [-f]\n",
                program_version);
        exit(1);
    }
  }

  /* Make sure not to consume the next input option */
  if (cmdstr[0] == '-') {
    fprintf (stderr, "Option -c requires an argument.\n");
    exit(1);
  }

  if (cmdstr[0]) {
    for (index = optind; index < argc; index++) {
      cmdlen += strlen(argv[index]) + 1;
      if (cmdlen < CMDMAXLEN) {
        strcat(cmdstr, " ");
        strcat(cmdstr, argv[index]);
      } else {
        fprintf (stderr, "Command too large. Ignoring.\n");
        cmdlen = 0;
        break;
      }
    }
  }

  if (cmdlen)
    fprintf(stdout, "cmd: [%s]\n", cmdstr);

  errno=0;  pw= getpwuid(getuid());
  if (!pw) { perror("password entry for uid not found"); exit(1); }
#ifdef SHADOW_PWD
  sp = getspnam(pw->pw_name);
  if (sp)
    pw->pw_passwd = sp->sp_pwdp;
  endspent();
#endif

  /* logically, if we need to do the following then the same
     applies to being installed setgid shadow.
     we do this first, because of a bug in linux. --jdamery */
  if (setgid(getgid())) { perror("setgid"); exit(1); }
  /* we can be installed setuid root to support shadow passwords,
     and we don't need root privileges any longer.  --marekm */
  if (setuid(getuid())) { perror("setuid"); exit(1); }

  if (strlen(pw->pw_passwd) < 13) {
    fputs("password entry has no pwd\n",stderr); exit(1);
  }

  display= XOpenDisplay(0);

  if (display==NULL) {
    fprintf(stderr,"xtrlock (version %s): cannot open display\n",
	    program_version);
    exit(1);
  }

  attrib.override_redirect= True;

  if (blank) {
    screen = DefaultScreen(display);
    attrib.background_pixel = BlackPixel(display, screen);
    window= XCreateWindow(display,DefaultRootWindow(display),
                          0,0,DisplayWidth(display, screen),DisplayHeight(display, screen),
                          0,DefaultDepth(display, screen), CopyFromParent, DefaultVisual(display, screen),
                          CWOverrideRedirect|CWBackPixel,&attrib);
    XAllocNamedColor(display, DefaultColormap(display, screen), "black", &black, &dummy);
  } else {
    window= XCreateWindow(display,DefaultRootWindow(display),
                          0,0,1,1,0,CopyFromParent,InputOnly,CopyFromParent,
                          CWOverrideRedirect,&attrib);
  }

  XSelectInput(display,window,KeyPressMask|KeyReleaseMask);

  csr_source= XCreateBitmapFromData(display,window,lock_bits,lock_width,lock_height);
  csr_mask= XCreateBitmapFromData(display,window,mask_bits,mask_width,mask_height);

  ret = XAllocNamedColor(display,
                        DefaultColormap(display, DefaultScreen(display)),
                        "steelblue3",
                        &dummy, &csr_bg);
  if (ret==0)
    XAllocNamedColor(display,
                    DefaultColormap(display, DefaultScreen(display)),
                    "black",
                    &dummy, &csr_bg);

  ret = XAllocNamedColor(display,
                        DefaultColormap(display,DefaultScreen(display)),
                        "grey25",
                        &dummy, &csr_fg);
  if (ret==0)
    XAllocNamedColor(display,
                    DefaultColormap(display, DefaultScreen(display)),
                    "white",
                    &dummy, &csr_bg);



  cursor= XCreatePixmapCursor(display,csr_source,csr_mask,&csr_fg,&csr_bg,
                              lock_x_hot,lock_y_hot);

  XMapWindow(display,window);

  /*Sometimes the WM doesn't ungrab the keyboard quickly enough if
   *launching xtrlock from a keystroke shortcut, meaning xtrlock fails
   *to start We deal with this by waiting (up to 100 times) for 10,000
   *microsecs and trying to grab each time. If we still fail
   *(i.e. after 1s in total), then give up, and emit an error
   */

  gs=0; /*gs==grab successful*/
  for (tvt=0 ; tvt<100; tvt++) {
    ret = XGrabKeyboard(display,window,False,GrabModeAsync,GrabModeAsync,
			CurrentTime);
    if (ret == GrabSuccess) {
      gs=1;
      break;
    }
    /*grab failed; wait .01s*/
    tv.tv_sec=0;
    tv.tv_usec=10000;
    select(1,NULL,NULL,NULL,&tv);
  }
  if (gs==0){
    fprintf(stderr,"xtrlock (version %s): cannot grab keyboard\n",
	    program_version);
    exit(1);
  }

  if (XGrabPointer(display,window,False,(KeyPressMask|KeyReleaseMask)&0,
               GrabModeAsync,GrabModeAsync,None,
               cursor,CurrentTime)!=GrabSuccess) {
    XUngrabKeyboard(display,CurrentTime);
    fprintf(stderr,"xtrlock (version %s): cannot grab pointer\n",
	    program_version);
    exit(1);
  }

  if (fork_after) {
    pid_t pid = fork();
    if (pid < 0) {
      fprintf(stderr,"xtrlock (version %s): cannot fork: %s\n",
              program_version, strerror(errno));
      exit(1);
    } else if (pid > 0) {
      exit(0);
    }
  }

  for (;;) {
    XNextEvent(display,&ev);
    switch (ev.type) {
    case KeyPress:
      if (ev.xkey.time < timeout) { XBell(display,0); break; }
      clen= XLookupString(&ev.xkey,cbuf,9,&ks,0);
      switch (ks) {
      case XK_Escape: case XK_Clear:
        rlen=0; break;
      case XK_Delete: case XK_BackSpace:
        if (rlen>0) rlen--;
        break;
      case XK_Linefeed: case XK_Return:
        if (rlen==0) break;
        rbuf[rlen]=0;
        if (passwordok(rbuf)) goto loop_x;
        XBell(display,0);
        rlen= 0;
        if (timeout) {
          goodwill+= ev.xkey.time - timeout;
          if (goodwill > MAXGOODWILL) {
            goodwill= MAXGOODWILL;
          }
        }
        timeout= -goodwill*GOODWILLPORTION;
        goodwill+= timeout;
        timeout+= ev.xkey.time + TIMEOUTPERATTEMPT;
        break;
      default:
        if (clen != 1) break;
        /* allow space for the trailing \0 */
	if (rlen < (sizeof(rbuf) - 1)){
	  rbuf[rlen]=cbuf[0];
	  rlen++;
	}
        break;
      }
      break;
    default:
      break;
    }
  }
 loop_x:
  if (cmdlen)
    execl("/bin/sh", "sh", "-c", cmdstr, (char *) 0);

  exit(0);
}
